src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs

```diff
@@ -77,26 +77,36 @@ private void ValidateScopeInput(ISet<string> scopesToValidate)
+            MsalTelemetryEventDetails telemetryEventDetails = null;
+            ITelemetryClient[] telemetryClients = null;
 
-            ApiEvent apiEvent = InitializeApiEvent(AuthenticationRequestParameters.Account?.HomeAccountId?.Identifier);
-            AuthenticationRequestParameters.RequestContext.ApiEvent = apiEvent;
-            MsalTelemetryEventDetails telemetryEventDetails = new MsalTelemetryEventDetails(TelemetryConstants.AcquireTokenEventName);
-            ITelemetryClient[] telemetryClients = AuthenticationRequestParameters.RequestContext.ServiceBundle.Config.TelemetryClients;
+            var measureTelemetryDurationResult = StopWatchService.MeasureCodeBlock(() =>
+            {
+                apiEvent = InitializeApiEvent(AuthenticationRequestParameters.Account?.HomeAccountId?.Identifier);
+                AuthenticationRequestParameters.RequestContext.ApiEvent = apiEvent;
+                telemetryEventDetails = new MsalTelemetryEventDetails(TelemetryConstants.AcquireTokenEventName);
+                telemetryClients = AuthenticationRequestParameters.RequestContext.ServiceBundle.Config.TelemetryClients;
+            });
 
             using (AuthenticationRequestParameters.RequestContext.CreateTelemetryHelper(apiEvent))
             {
                 try
                 {
-                    AuthenticationRequestParameters.LogParameters();
-                    LogRequestStarted(AuthenticationRequestParameters);
+                    AuthenticationResult authenticationResult = null;
+                    var measureDurationResult = await StopWatchService.MeasureCodeBlockAsync(async () =>
+                    {
+                        AuthenticationRequestParameters.LogParameters();
+                        LogRequestStarted(AuthenticationRequestParameters);
+
+                        authenticationResult = await ExecuteAsync(cancellationToken).ConfigureAwait(false);
+                        LogReturnedToken(authenticationResult);
+                    }).ConfigureAwait(false);
 
-                    AuthenticationResult authenticationResult = await ExecuteAsync(cancellationToken).ConfigureAwait(false);
-                    LogReturnedToken(authenticationResult);
-                    UpdateTelemetry(sw, apiEvent, authenticationResult);
+                    UpdateTelemetry(measureDurationResult.Milliseconds + measureTelemetryDurationResult.Milliseconds, apiEvent, authenticationResult);
                     LogMetricsFromAuthResult(authenticationResult, AuthenticationRequestParameters.RequestContext.Logger);
                     LogSuccessfulTelemetryToClient(authenticationResult, telemetryEventDetails, telemetryClients);
-                    LogMsalSuccessTelemetryToOtel(authenticationResult, apiEvent.ApiId.ToString(), sw.ElapsedTicks / (TimeSpan.TicksPerMillisecond / 1000));
+                    LogMsalSuccessTelemetryToOtel(authenticationResult, apiEvent.ApiId.ToString(), measureDurationResult.Milliseconds * 1000);
```

Comment:
This would not be an accurate measurement in microseconds. The reason we are recording the duration in microseconds is because when a cache is used, many times the duration is 0 ms. So to get a more accurate duration we are recording in microseconds which is calculated using ticks. In this scenario, when the duration in ms is 0, it will be 0 in microseconds as well.

Reply:
I would recommend capturing duration in ticks instead of milliseconds. https://identitydivision.visualstudio.com/DevEx/_git/MISE?path=/src/Caching/Utility.cs&version=GBgeok/telemetry-exporter&line=30&lineEnd=31&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents